gridsync/streamedlogs.py

```diff
@@ -0,0 +1,98 @@
+# -*- coding: utf-8 -*-
+
+"""
+Support for reading the streaming Eliot logs available from a Tahoe-LAFS
+node.
+"""
+
+import logging
+from collections import deque
+from urllib.parse import urlsplit
+
+from hyperlink import parse
+
+from twisted.internet.endpoints import TCP4ClientEndpoint
+from twisted.application.internet import ClientService
+from twisted.application.service import MultiService
+
+from autobahn.twisted.websocket import (
+    WebSocketClientFactory,
+    WebSocketClientProtocol,
+)
+
+
+class TahoeLogReader(WebSocketClientProtocol):
+    def onMessage(self, payload, isBinary):
+        if isBinary:
+            logging.warning(
+                "Received a binary-mode WebSocket message from Tahoe-LAFS "
+                "streaming log; dropping.",
+            )
+        else:
+            self.factory.streamedlogs.add_message(payload)
+
+
+class StreamedLogs(MultiService):
+    """
+    :ivar _reactor: A reactor that can connect using whatever transport the
+        Tahoe-LAFS node requires (TCP, etc).
+
+    :ivar deque _buffer: Bounded storage for the streamed messages.
+    """
+    _started = False
+
+    def __init__(self, reactor, maxlen=None):
+        super().__init__()
+        self._reactor = reactor
+        if maxlen is None:
+            # This deque limit is based on average message size of 260 bytes
+            # and a desire to limit maximum memory consumption here to around
+            # 500 MiB.
+            maxlen = 2000000
```

Comment:
Something that occurs to me is that some folks use GridSync with multiple grids.  Do we want to divide this buffer across all of the Tahoe's being monitored?

cc @crwood

Reply:
I was actually just thinking about this last night.. Longer (or medium?) term I'd like to have this value and behavior be somewhat user-configurable via the Preferences pane (and/or forthcoming "Connection Manager"), however, in the interest of avoiding further feature-creep for _this_ PR (i.e., keeping things focused primarily on _consuming_ the logs), I'm happy to merge this work as-is and implement the necessary memory-balancing dials later on.

Nevertheless, I _may_ lower this value somewhat pending further testing once I hook this work into the GUI that I'm working on in [`168.export-debug-info`](https://github.com/gridsync/gridsync/tree/168.export-debug-info) (as it presents both "filtered" and "unfiltered" versions of logs to users and so requires additional memory).